### PR TITLE
Upgrade to .NET 8.0

### LINKED ISSUE #401: [Feature Contribution]: Upgrade .NET version to .NET 8
### What happened?

.NET 8 that is the next LTS version of .NET is planned to be release Nov. 2023.  Carbon Aware SDK should be upgraded to .NET 8 because using the latest LTS is good practice, generally. 

FYI, .NET 8 contains important improvements to implement #400. 
* https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#metrics
* https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#imeterfactory-interface

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Feature Commitment

- [X] I commit to contributing this feature as a PR and working with the GSF to merge this feature into the Carbon Aware SDK.

### PR DESCRIPTION
**Pull request by tiwatsuka**
# Pull Request

Issue Number:  #401 

## Summary

Upgrade .NET version from 6.0 to 8.0.

## Changes

- change value of `<TargetFramework>` in each `.csproj` file
- change base images of containers for development and building
- update .NET version contained  in comments, documents, and settings

## Checklist

- [x] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [x] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [x] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?

No.

## Is this a breaking change?

No.

This PR Closes #401

### CODE REVIEW

**Review by bderusha — APPROVED** [review]
These changes look reasonable. If the tests all pass :shipit:

**Comment by Willmish** [comment]
@tiwatsuka highlighted https://techcommunity.microsoft.com/t5/apps-on-azure-blog/net-on-azure-functions-august-2023-roadmap-update/ba-p/3910098 :
- Current Sample Azure Functions uses In process model rather than Isolated worker model 
image

**Comment by Sophietn** [comment]
This is blocked until we update Sample Azure Functions from In process model to Isolated worker model - this will be done in a new issue

**Comment by Willmish** [comment]
To pass https://github.com/Green-Software-Foundation/carbon-aware-sdk/actions/runs/6879323122/job/18711361618?pr=404 , this requires a manual check, with proof of  api-comparison passing (this is currently trying to compare to non-existing previous dotnet8 swagger definition)

**Review by YaSuenag — CHANGES_REQUESTED** [review]
Please update [containerization.md](https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/c0e56baa02574a47b5af1e9d7c9f88dada0f62f8/docs/containerization.md#run-webapi-image) because it mentions exporsed port number of WebAPI. It should be `8080`.

https://learn.microsoft.com/en-us/dotnet/core/compatibility/containers/8.0/aspnet-port

File: helm-chart/templates/deployment.yaml
Comment (YaSuenag) [inline helm-chart/templates/deployment.yaml:1]: You've changed `containerPort` which cannot be modified via values.yaml, so we have to publish new chart. Could you bump up chart version like this?

https://github.com/Green-Software-Foundation/carbon-aware-sdk/pull/449/commits/98c45436744ef04d1d9e8b652ae7de5ec4a1b759

Note that CASDK maintainer should kick [5-publish-helm-chart.yaml](https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/dev/.github/workflows/5-publish-helm-chart.yaml) publish after merging.

Comment (YaSuenag) [reply helm-chart/templates/deployment.yaml:1]: I think `1.1.1` is a good version for this.

Comment (YaSuenag) [reply helm-chart/templates/deployment.yaml:1]: I reconsidered that we have to release new chart after `vNext` because the chart includes application version (it is equivalent to the tag of container image).

So we can bump up chart version in another PR after `vNext`.

**Review by YaSuenag — APPROVED** [review]
LGTM

Thanks for your work!

**Comment by YaSuenag** [comment]
@vaughanknight @danuw I agree this PR.

**Comment by tiwatsuka** [comment]
@danuw I've just resolved documentation conflicts. 
The workflows are not executed because they runs only for PR to `main` or `dev`. Could you execute them manually?

**Comment by YaSuenag** [comment]
I think we can add `vnext/dotnet8-upgrade` to `.on.pull_request.branches` in `1-pr.yaml` in this branch temporally to run PR tests. Maybe it can run on draft PR, so you can move this to draft PR under PR testing, and move to PR again after removing this branch from `1-pr.yaml` when all tests has been passed. Does it make sence?

**Comment by tiwatsuka** [comment]
All checks have passed except `api-comparison`. 
https://github.com/Green-Software-Foundation/carbon-aware-sdk/actions/runs/8081752762

`api-comparison` cannot be checked this PR automatically as mentioned before.
https://github.com/Green-Software-Foundation/carbon-aware-sdk/pull/404#issuecomment-1829326957

**Comment by YaSuenag** [comment]
It's good news!

@vaughanknight @danuw Can you merge this to vnext/dotnet8-upgrade branch? #422 relies on this PR - it needs to be rebase after this PR.

**Review by vaughanknight — COMMENTED** [review]
Looks good as discussed on the call, and all conflicts resolved.  Thanks all!

**Pull request merged**
